cartographer/mapping_2d/local_trajectory_builder.cc

```diff
@@ -121,26 +121,29 @@ LocalTrajectoryBuilder::AddRangeData(const common::Time time,
 
   if (num_accumulated_ >= options_.scans_per_accumulation()) {
     num_accumulated_ = 0;
-    return AddAccumulatedRangeData(
-        time, sensor::TransformRangeData(accumulated_range_data_,
-                                         tracking_delta.inverse()));
+    // Transforms 'accumulated_range_data_' into a frame where gravity direction
+    // is approximately +z.
+    const transform::Rigid3d gravity_alignment = transform::Rigid3d::Rotation(
+        extrapolator_->EstimateGravityOrientation(time));
+    const sensor::RangeData gravity_aligned_range_data =
+        TransformAndFilterRangeData(
+            gravity_alignment.cast<float>() * tracking_delta.inverse(),
+            accumulated_range_data_);
+    if (gravity_aligned_range_data.returns.empty()) {
```

Comment:
Move this check back to `AddAccumulatedRangeData()` and you can get rid of the temporary. What do you think?